src/app/services/baw-api/resolver-common.ts

```diff
@@ -1,53 +1,124 @@
-import { ActivatedRouteSnapshot, ParamMap, Resolve } from "@angular/router";
+import { Type } from "@angular/core";
+import { ActivatedRouteSnapshot, Resolve } from "@angular/router";
 import { Observable, of } from "rxjs";
 import { catchError, map, take } from "rxjs/operators";
 import { Id } from "src/app/interfaces/apiInterfaces";
 import { AbstractModel } from "src/app/models/AbstractModel";
-import { ApiShow, IdOr, ApiList } from "./api-common";
+import { ApiList, ApiShow, IdOr } from "./api-common";
 import { ApiErrorDetails } from "./api.interceptor.service";
 
-export class ListResolver<T extends AbstractModel>
-  implements Resolve<ResolvedModel<T[]>> {
+export class Resolvers<
+  T extends AbstractModel,
+  A extends ApiList<T, any[]> & ApiShow<T, any[], IdOr<T>>
+> {
   constructor(
-    private api: ApiList<T, any[]>,
-    private ids: (params: ParamMap) => Id[]
+    private deps: Type<A>[],
+    private id?: string,
+    private ids?: string[]
   ) {}
 
-  public resolve(
-    route: ActivatedRouteSnapshot
-  ): Observable<ResolvedModel<T[]>> {
-    return this.api.list(...this.ids(route.paramMap)).pipe(
-      map(model => ({ model })),
-      take(1),
-      catchError((error: ApiErrorDetails) => {
-        return of({ error });
-      })
-    );
+  public create(name: string) {
+    const deps = this.deps;
+    const id = this.id;
+    const ids = this.ids;
+
+    const listProvider = new ListResolver<T, A>(deps, ids).create(name);
+    const showProvider = new ShowResolver<T, A>(deps, id, ids).create(name);
+
+    return [...listProvider, ...showProvider];
+  }
+}
+
+export class ListResolver<
+  T extends AbstractModel,
+  L extends ApiList<T, any[]>
+> {
+  constructor(private deps: Type<L>[], private ids?: string[]) {}
+
+  public create(name: string) {
+    const deps = this.deps;
+    const ids = this.ids;
+
+    class Resolver implements Resolve<ResolvedModel<T[]>> {
+      constructor(private api: L) {}
+
+      public resolve(
+        route: ActivatedRouteSnapshot
+      ): Observable<ResolvedModel<T[]>> {
+        const args: Id[] = ids
+          ? ids.map(id => convertToId(route.paramMap.get(id)))
+          : [];
+
+        return this.api.list(...args).pipe(
+          map(model => ({ model })),
+          take(1),
+          catchError((error: ApiErrorDetails) => {
+            return of({ error });
+          })
+        );
+      }
+    }
+
+    return [
+      {
+        provide: name + "sResolver",
```

Comment:
I wonder if you could use `typeof` here to get typescript to put the name is for us?

Also, don't bother with the pluralization... it's not worth it in the end. Why not simpler: `XXXListResolver`

Reply:
Unfortunately `typeof` does not work on custom classes, only `instanceof`